test/Ocelot.UnitTests/Configuration/UpstreamTemplatePatternCreatorTests.cs

```diff
@@ -233,7 +234,23 @@ public void should_create_template_pattern_that_matches_query_string_with_multip
                 .Then(x => x.ThenTheFollowingIsReturned($"^(?i)/api/subscriptions/[^/]+/updates(|\\?)unitId={MatchEverything}&productId={MatchEverything}$"))
                 .And(x => ThenThePriorityIs(1))
                 .BDDfy();
-        }
+        }
+
+        [Fact]
+        public void should_fail_gracefully_when_placeholder_name_is_duplicate()

```

Comment:
On what app stage we have the failure?
During route validation stage?
After validation, during `DownstreamRoute` object creation?
Another stage?

Reply:
@raman-m at replacing placeholders with the correct regex sequence

Reply:
This is quite too late stage ))
Please, keep this exception while replacing placeholders by their values. And write `Error` event into the log. I know this exception will be caught by Error Mapper. So, probably we will have 2 records in the log. Need testing this...

Write `Critical` event into the log during validation stage.

Reply:
@raman-m regarding this part, the route validater builds the route first before validating it, so even if I try to handle this in the route validation it will be even later, is there any earlier stages I should look into, If I try to handle this also at configuration builder stage I will have to duplicate the logic of parsing the placeholder from different places for example the upstreampatterncreator.cs

Reply:
Let's make final decision...
I believe `UpstreamTemplatePatternCreator` is not a right place to check the placeholders. Of course, you catch the problem even earlier than validators, but the issue is mixing responsibilities 👉 
- `UpstreamTemplatePatternCreator` creates a template via replacing template string with converted/parsed regular expression.
- Validators run checking functions via attached validation rules

```diff
! So, we have the violation of Single Responsibility principle here, aka SOLID
```
Please move checking of placeholders uniqueness to a validator please, and define a rule, and remove this exception for sure from `UpstreamTemplatePatternCreator` class. Thanks!

Reply:
@raman-m Ok I understand your point, but the upstreamtemplate is built before the validator rules are checked, so even if I add the rule in RouteFluentValidator.cs the exception will still happen. What should I do in that part ?

Reply:
Yes, the `RouteFluentValidator` class is the right place.
Please remove exception! Seems I've explained you why... **`UpstreamTemplatePatternCreator`  should have zero changes!**

I meant this live
https://github.com/ThreeMammals/Ocelot/pull/1927/files#diff-611108f6b3268be0a8735448eb7b2b23e994a1bbf18b4a99794a9f720cdd121aR40

Reply:
It seems this has already been fixed. We've changed design to validate route JSON during Ocelot starting.
So, no exceptions in Ocelot kernel now. Good! 👍